### PR TITLE
CommandAudit and SessionAudit support

### LINKED ISSUE #2340: tpm2_create: cpHash was calculated for the wrong command code
When tpm2_create is passed the option to calculate a template, it should ideally invoke TPM2_CC_CreateLoaded.
There is a bug that causes TPM2_CC_Create to be invoked instead.

### LINKED ISSUE #2341: tpm2_create: cpHash algorithm is incorrectly chosen from the first session handle passed
Ideally, The cpHash algorithm should be chosen in the following order:

If cphash_path is preceded with : use that as cphash-halg return
Otherwise
Consume session only if it is a policy-session or an hmac-session

If only hmac or policy session is specified, return that session's halg
If hmac-session with audit is specified, return that session's halg
If policy-session, then return policy-session's halg
Otherwise
return SHA256

### LINKED ISSUE #2342: tpm2_create: template outputs are prohibited unless the TPM2_CC_<command> is executed
Ideally, template output can also be executed without dispatching the TPM2_CC_Create/loaded command.

### LINKED ISSUE #2343: tpm2_create: When template option is given, the inPublic data should be that of TPM2_CC_CreateLoaded
When calculating the template data, there is a bug that the inPublic data was coming through the initialization of TPM2_CC_Create instead of TPM2_CC_CreateLoaded.

### LINKED ISSUE #2347: tpm2_activatecredential: Add aux session support
Add support for specifying non-auth sessions for auditing and parameter encryption.

### LINKED ISSUE #2348: tools/tpm2_activatecredential: Add support for deducing cphash alg using lib/tpm2_util_calculate_phash_alg

### LINKED ISSUE #2349: tools/tpm2_activatecredential: Add support for calculating rpHash for sessiondigests

### LINKED ISSUE #2350: tpm2_certify: Add auxiliary session support

### LINKED ISSUE #2351: tpm2_certify: Add support for deducing cphash alg using lib/tpm2_util_calculate_phash_alg

### LINKED ISSUE #2352: tpm2_certify: Add support for calculating rpHash for sessiondigests

### LINKED ISSUE #2356: tpm2_certifycreation: Add support for deducing cphash alg using lib/tpm2_util_calculate_phash_alg

### LINKED ISSUE #2357: tpm2_certifycreation: Add support for calculating rpHash for sessiondigests

### LINKED ISSUE #2358: tpm2_certifycreation:  Add auxiliary session support

### LINKED ISSUE #2364: tpm2_changeauth: Add support for deducing cphash alg using lib/tpm2_util_calculate_phash_alg

### LINKED ISSUE #2365: tpm2_changeauth: Add support for calculating rpHash for sessiondigests

### LINKED ISSUE #2366: tpm2_changeauth: Add auxiliary session support

### LINKED ISSUE #2369: tpm2_changeeps: Add cpHash support

### LINKED ISSUE #2370: tpm2_changeeps: add support for calculating rpHash for sessiondigests

### LINKED ISSUE #2371: tpm2_changeeps: add auxiliary session support

### LINKED ISSUE #2372: tpm2_changepps: Add cpHash support

### LINKED ISSUE #2373: tpm2_changepps: add support for calculating rpHash for sessiondigests

### LINKED ISSUE #2374: tpm2_changepps: add auxiliary session support

### PR DESCRIPTION
**Pull request by idesai**
Fixes #2372
Fixes #2373
Fixes #2374
Fixes #2371
Fixes #2370
Fixes #2369
Fixes #2366
Fixes #2365
Fixes #2364 
Fixes #2358
Fixes #2357
Fixes #2356 
Fixes #2350 
Fixes #2352 
Fixes #2351 
Fixes #2349
Fixes #2347
Fixes #2348 
Fixes #2340 
Fixes #2341
Fixes #2342
Fixes #2343

It should be noted that the auths aren't checked when calculating the cpHash,
instead the sessions are consumed to determine the cpHash algorithm.

1. If cphash_path is preceded with <halg>: use that as cphash-halg return
Otherwise

 Consume session only if it is a policy-session or an hmac-session
 1. If only hmac or policy session is specified, return that session's halg
 2. If hmac-session with audit is specified, return that session's halg
 3. If policy-session, then return policy-session's halg

Otherwise
  return SHA256

Signed-off-by: Imran Desai <imran.desai@intel.com>

### CODE REVIEW

File: lib/tpm2_util.c [hunk 1]
```diff
@@ -1100,11 +1102,142 @@ TPMI_ALG_HASH *param_hash_algorithm, tpm2_session **session) {
...
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        rethash = tpm2_session_get_authhash(session);
+
+        TPMA_SESSION attrs = 0;
+        ESYS_TR session_handle = tpm2_session_get_handle(session);
+        tool_rc tmp_rc = tpm2_sess_get_attributes(ectx, session_handle, &attrs);
+        UNUSED(tmp_rc);
```
Comment (williamcroberts) [inline lib/tpm2_util.c:1172]: this is okay to Fail?

File: lib/tpm2_util.c [hunk 2]
```diff
@@ -1100,11 +1102,142 @@ TPMI_ALG_HASH *param_hash_algorithm, tpm2_session **session) {
...
+    TPMI_ALG_HASH cphash_alg = calc_cphash_alg_from_cphash_path(cphash_path);
+    if (cphash_alg != TPM2_ALG_ERROR) {
+        return cphash_alg;
+    }
+
+    bool is_sessions_path_valid = adjust_sessions_auths_or_paths(
+        sessions_auths_or_paths);
+    if (!is_sessions_path_valid) {
+        LOG_WARN("Invalid session count or paths: Defaulting cphash alg to sha256");
+        return TPM2_ALG_SHA256;
```
Comment (williamcroberts) [inline lib/tpm2_util.c:1233]: Is this user input? If user specifies input, shouldn't we fail?

File: lib/tpm2_util.c [hunk 3]
```diff
@@ -1100,11 +1102,142 @@ TPMI_ALG_HASH *param_hash_algorithm, tpm2_session **session) {
...
+ * 1. If cphash_path is preceded with <halg>: use that as cphash-halg return
+ * Otherwise
+ *
+ *  Consume session only if it is a policy-session or an hmac-session
+ *  1. If only hmac or policy session is specified, return that session's halg
+ *  2. If hmac-session with audit is specified, return that session's halg
+ *  3. If policy-session, then return policy-session's halg
+ *
+ * Otherwise
+ *  return SHA256
```
Comment (williamcroberts) [inline lib/tpm2_util.c:1219]: Whats the otherwise case(s)?

**Review by williamcroberts — COMMENTED** [review]
Where are your tests?

File: lib/tpm2_util.c (hunk 3 above)
Comment (idesai) [reply lib/tpm2_util.c:1219]: Basically, password sessions and some commands like getrandom that don't require an auth-session at all.

File: lib/tpm2_util.c (hunk 1 above)
Comment (idesai) [reply lib/tpm2_util.c:1172]: Correct. If that happens we return the default SHA256.

File: lib/tpm2_util.c (hunk 2 above)
Comment (idesai) [reply lib/tpm2_util.c:1233]: I guess is_sessions_path_valid could be better phrased and also remove the warning to make the outputs less noisy :). It is possible that user correctly specifies password-only sessions and in those cases we default to sha256.

**Comment by williamcroberts** [comment]
The CI failure seems to be affecting even PRs that don't make code changes like #2346. I would assume this a memory issue, or something with clang-7 that we bumped into. I set my cpu's to 2 and my my memory to 14GB in my docker container to try and more closely replicate the hardware of Github Actions:
  - https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

**I was able to replicate it** I added the following to my docker run command:
`--memory=14GB --cpus=2`

I ran through clang-7, and clang-9. It seems clang-t is still affected, but clang-9 is OK.

Rather than wasting our time, clang-9 seems to avoid this. I'll work on updating our images tonight, so tomorrow we can resolve these issues.

**Comment by williamcroberts** [comment]
@idesai I updated ubuntu-18.04 image, i retriggered, but now I am seeing:
```
../../tools/tpm2_activatecredential.c: In function 'tpm2_tool_onrun':
../../tools/tpm2_activatecredential.c:208:67: error: array subscript 2 is above array bounds of 'const char *[1]' [-Werror=array-bounds]
  208 |             all_sessions_paths[session_idx] = ctx.aux_session_path[session_idx];
      |                                               ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

But it was building on CC=clang for ubuntu-18.04 with the branch @ 6df5fa7

**Comment by idesai** [comment]
> @idesai I updated ubuntu-18.04 image, i retriggered, but now I am seeing:
> 
> ```
> ../../tools/tpm2_activatecredential.c: In function 'tpm2_tool_onrun':
> ../../tools/tpm2_activatecredential.c:208:67: error: array subscript 2 is above array bounds of 'const char *[1]' [-Werror=array-bounds]
>   208 |             all_sessions_paths[session_idx] = ctx.aux_session_path[session_idx];
>       |                                               ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> ```
> 
> But it was building on CC=clang for ubuntu-18.04 with the branch @ [6df5fa7](https://github.com/tpm2-software/tpm2-tools/commit/6df5fa79a216145d7381b61f5d8b7bb0686b6ebc)

@williamcroberts, unfortunately none of the fapi tests are passing because they cannot access the files for some reason.

**Comment by williamcroberts** [comment]
Nice. I think you officially merged the longest list of fixes ever.

**Pull request merged**